### PR TITLE
(#4595) -  validate urls in replicate.from and replicate.to

### PR DESCRIPTION
**Pull request by daedlock**

### CODE REVIEW

**Comment by daleharvey** [comment]
Thanks for the PR :) the setup looks great however I am not certain what we want to do is validate urls here, validating urls is an extremely tricky business and the url here isnt exhaustive, I think what we should be doing is ensuring any error thrown by ajax is properly returned back to the user

Currently it looks like replicate.to will not either complete or throw when hitting this error, so a valid test can look like 

```
    new PouchDB(dbs.name).replicate.to('http:///').then(function() { 
      done();
    }).catch(function(e) {
      done();
    });
```

**Comment by daleharvey** [comment]
Talked to @daedlock on irc and said they are working on changing the PR to avoid reporting error, but pinging @nolanlawson just to check you agree since we didnt say the same thing on the original issue. If we were to do url testing then it should likely go in the http adapter during construction, but I think we want to avoid that really, url parsing is complicated and we have specific features to override ajax so that users dont even need to provide valid urls. I think its best we make sure we propagate the error correctly.

**Comment by daedlock** [comment]
As discussed with @daleharvey. The problem with propagating the error is `NS_MALFORMED_URL` gets thrown earlier in Firefox, so I think the only solution here is to do some sort of a obvious url validation like checking for illegal characters, number of slashes..etc. The validation checks make sense to be done in the `replicate()` constructor and if it fails it would reject the promise and emit an error. So basically I am talking about a wrapper that detects the problem before it happens.

**Comment by daleharvey** [comment]
Sorry little miscommunication, I dont think we should be doing pre validation, I think whatever line throws the malformed url error (I am guessing xhr.send @ https://github.com/pouchdb/pouchdb/blob/master/lib/deps/ajax/request-browser.js#L199) we should catch that error and propogate it back to the user. 

It should hopefully be the matter of a try / catch around the send and calling `callback(err)` when caught

File: lib/deps/ajax/request-browser.js [hunk 1]
```diff
@@ -120,7 +120,13 @@ function xhRequest(options, callback) {
     xhr = new XMLHttpRequest();
   }
 
-  xhr.open(options.method, options.url);
+  try {
+    xhr.open(options.method, options.url);
+  }
+  catch(exception) {
```
Comment (daleharvey) [inline lib/deps/ajax/request-browser.js]: nit: We put the catch on the same line generally
``` 
} catch (exception) { 
```

File: tests/integration/test.basics.js [hunk 2]
```diff
@@ -83,6 +83,22 @@ adapters.forEach(function (adapter) {
       });
     });
 
+    it('[4595] should reject xhr errors', function(done){
+      var invalidUrl = 'http:///';
+      new PouchDB(dbs.name).replicate.to(invalidUrl,{})
+      .catch(function(err,changes){
+        done();
+      });
```
Comment (daleharvey) [inline tests/integration/test.basics.js]: nit: this doesnt quite match our code style, should look as follows
```
new PouchDB(dbs.name).replicate.to('http:///').catch(function() {
  done();
});
```

**Comment by daleharvey** [comment]
There are tiny nits on the code style, but I think this looks great, can edit the code style issues on the way in, lets just wait to see if the tests are green

File: tests/integration/test.basics.js (hunk 2 above)
Comment (daedlock) [inline tests/integration/test.basics.js]: It wont pass  jshint "too long line". Should we ignore?

**Comment by nolanlawson** [comment]
Sweet, tests are green! By the way, I have a huge PR coming in #4710, but I'll consider it my responsibility to fix merge conflicts.

File: tests/integration/test.basics.js (hunk 2 above)
Comment (nolanlawson) [inline tests/integration/test.basics.js]: hm, it does if you remove the args from the function (which is fine)

**Comment by nolanlawson** [comment]
Fixed the nits and merged as d3aa2c92b0a5bf461384184211bde68ae4037eec. Nice work!

**Pull request closed without merging**